### PR TITLE
Kyc level check. 

### PR DESCRIPTION
**Pull request by MoMannn**
Logic: check the level of the last kycToken the msg.sender got.

### CODE REVIEW

File: contracts/crowdsale/ZxcCrowdsale.sol [hunk 1]
```diff
@@ -204,9 +204,13 @@ contract ZxcCrowdsale is
     uint256 tokens;
 
     // Sender needs Xcert KYC token.
-    require(xcertKyc.balanceOf(beneficiary) > 0);
-
+    uint256 balance = xcertKyc.balanceOf(msg.sender);
+    require(balance > 0);
+    
     if (isPresale()) {
+      uint256 tokenId = xcertKyc.tokenOfOwnerByIndex(beneficiary, balance.sub(1));
```
Comment (lknix) [inline contracts/crowdsale/ZxcCrowdsale.sol]: We do expect for every user to just have one KYC token, right? Maybe we should be more strict - balance should be == 1 and token index is then 0.

File: test/crowdsale/ZxcCrowdsale.test.js [hunk 2]
```diff
@@ -40,11 +41,15 @@ contract('crowdsale/ZxcCrowdsale', (accounts) => {
   const buyerTwo = accounts[5];
   const _tester = accounts[6];  // tester should never be the default account!
   const xcertTokenOwner = accounts[7];
+  const buyerThree = accounts[8];
 
   let token;
   let xcertToken;
   let crowdsale;
 
+  const config = [web3Util.padLeft(web3Util.numberToHex(0), 64)];
```
Comment (lknix) [inline test/crowdsale/ZxcCrowdsale.test.js:48]: I think we need a helper function in utils which is taking care of this (not a fix requested in this PR tho)

File: contracts/crowdsale/ZxcCrowdsale.sol (hunk 1 above)
Comment (MoMannn) [reply contracts/crowdsale/ZxcCrowdsale.sol]: Even tho we do expect that everyone will have only 1 KYC token there is a possibility that something unexpected happens. So this is just a precaution. 

@xpepermint What do you think?

File: test/crowdsale/ZxcCrowdsale.test.js [hunk 3]
```diff
@@ -733,6 +739,52 @@ contract('crowdsale/ZxcCrowdsale', (accounts) => {
       assert.strictEqual(actualTokens.toString(), expectedTokens.add(expectedBonus).toString());
     });
 
+    it('buyTokens should fail purchasing tokens if buyers kyc token is level 1 in presale', async () => {
```
Comment (lknix) [inline test/crowdsale/ZxcCrowdsale.test.js]: you need to update test descriptions for other presale tests - to include KYC level information

File: test/crowdsale/ZxcCrowdsale.test.js [hunk 4]
```diff
@@ -733,6 +739,52 @@ contract('crowdsale/ZxcCrowdsale', (accounts) => {
...
+                            "2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae",
+                            config,
+                            data,
+                            {from: xcertTokenOwner});
+     
+      await increaseTimeTo(startTimePresale + duration.seconds(30));
+      await assertRevert(crowdsale.buyTokens({from: buyerThree, value: weiAmount}));
+    });
+
+    it('buyTokens should purchase tokens if buyers has multiple kyc tokens but the last one is level 2 in presale', async () => {
```
Comment (lknix) [inline test/crowdsale/ZxcCrowdsale.test.js]: I wouldn't even support this

File: test/crowdsale/ZxcCrowdsale.test.js (hunk 2 above)
Comment (MoMannn) [reply test/crowdsale/ZxcCrowdsale.test.js:48]: Yeah a function like numberToBytes32 would be welcomed. 

@xpepermint Can we package something like this directly into chainspin?

File: test/crowdsale/ZxcCrowdsale.test.js [hunk 5]
```diff
@@ -744,6 +796,25 @@ contract('crowdsale/ZxcCrowdsale', (accounts) => {
       assert.strictEqual(actualTokens.toString(), expectedTokens.toString());
     });
 
+    it('buyTokens should purchase tokens if buyer kyc token is level 1 in crowdsale', async () => {
```
Comment (lknix) [inline test/crowdsale/ZxcCrowdsale.test.js]: Update existing test for buyerOne to mention KYC level

Comment (MoMannn) [reply test/crowdsale/ZxcCrowdsale.test.js]: https://github.com/0xcert/ethereum-crowdsale/pull/7/commits/090f7533070cb6977cd8918bd05b17c54b104447

File: test/crowdsale/ZxcCrowdsale.test.js (hunk 3 above)
Comment (MoMannn) [reply test/crowdsale/ZxcCrowdsale.test.js]: https://github.com/0xcert/ethereum-crowdsale/pull/7/commits/090f7533070cb6977cd8918bd05b17c54b104447

File: test/crowdsale/ZxcCrowdsale.test.js [hunk 6]
```diff
@@ -580,328 +577,311 @@ contract('crowdsale/ZxcCrowdsale', (accounts) => {
...
+                              {from: xcertTokenOwner});
+      });
+
+      it('buyTokens should fail in presale', async () => {
+        const weiAmount = ether(1);
+        await increaseTimeTo(startTimePresale + duration.seconds(30));
+        await assertRevert(crowdsale.buyTokens({from: buyer, value: weiAmount}));
+      });
+  
+      it('buyTokens should purchase tokens in presale if another kyc level 2 token is minter after level 1 token', async () => {
```
Comment (lknix) [inline test/crowdsale/ZxcCrowdsale.test.js]: typo: minted

File: test/crowdsale/ZxcCrowdsale.test.js [hunk 7]
```diff
@@ -580,328 +577,311 @@ contract('crowdsale/ZxcCrowdsale', (accounts) => {
...
+        data = [web3Util.padLeft(web3Util.numberToHex(1), 64)];
+        await increaseTimeTo(startTimeSaleNoBonus + duration.seconds(30));
+        await crowdsale.buyTokens({from: buyer, value: weiAmount});
+        const actualTokens = await token.balanceOf(buyer);
+        assert.strictEqual(actualTokens.toString(), expectedTokens.toString());
+      });
+    });
+
+    describe('no KYC', function() {
+      it('buyTokens should fail', async () => {
```
Comment (lknix) [inline test/crowdsale/ZxcCrowdsale.test.js]: should fail if/when ...

Comment (lknix) [reply test/crowdsale/ZxcCrowdsale.test.js]: ah, ok it's "no KYC" printed prior to this

Comment (lknix) [reply test/crowdsale/ZxcCrowdsale.test.js]: but it should fail prior, during presale, crowdsale and after - tests

File: test/crowdsale/ZxcCrowdsale.test.js [hunk 8]
```diff
@@ -580,328 +577,311 @@ contract('crowdsale/ZxcCrowdsale', (accounts) => {
...
+        await xcertToken.mint(buyer,
+                              123,
+                              "https://foobar.io",
+                              "2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae",
+                              config,
+                              data,
+                              {from: xcertTokenOwner});
+      });
+
+      it('buyTokens should fail in presale', async () => {
```
Comment (lknix) [inline test/crowdsale/ZxcCrowdsale.test.js:834]: should fail prior, presale, after. Should succeed in crowdsale /w bonus and w/o bonus.

**Comment by lknix** [comment]
Let's add full coverage for buyTokens which is paramount.

File: test/crowdsale/ZxcCrowdsale.test.js [hunk 9]
```diff
@@ -869,7 +880,21 @@ contract('crowdsale/ZxcCrowdsale', (accounts) => {
...
+      });
+
+      it('buyTokens should fail in crowdsale with bonus', async () => {
+        const weiAmount = ether(2.1);
+  
+        await increaseTimeTo(startTimeSaleWithBonus + duration.seconds(30));
+        await assertRevert(crowdsale.buyTokens({from: buyer, value: weiAmount}));
+      });
+
+      it('buyTokens should fail in crowdsale', async () => {
```
Comment (lknix) [inline test/crowdsale/ZxcCrowdsale.test.js]: change to: 'buyTokens should fail in crowdsale with no bonus'

**Comment by lknix** [comment]
One last comment. When fixed -> LGTM

**Pull request merged**